virttest/qemu_devices/qcontainer.py

```diff
@@ -1718,6 +1719,145 @@ def images_define_by_params(self, name, image_params, media=None,
+                                 file_name=None):
+        """
+        Wrapper for creating serial device from serial params.
+
+        :param serial_id: id of serial object
+        :param params: serial params
+        :param file_name: the file path of the serial device (optional)
+        :return: the device list that construct the serial device
+        """
+        serial_type = params['serial_type']
+        machine = params.get('machine_type')
+
+        # Arm lists "isa-serial" as supported but can't use it,
+        # fallback to "-serial"
+        legacy_cmd = " -serial unix:'%s',server,nowait" % file_name
+        legacy_dev = qdevices.QStringDevice('SER-%s' % serial_id,
+                                            cmdline=legacy_cmd)
+        arm_serial = (serial_type == 'isa-serial'
+                      and 'arm' in params.get("machine_type", ""))
+        if (arm_serial or not self.has_option("chardev")
+                or not self.has_device(serial_type)):
+            return legacy_dev
+
+        bus_type = None
+        if serial_type.startswith('virt'):
+            if '-mmio' in machine:
+                controller_suffix = 'device'
+            elif machine.startswith("s390"):
+                controller_suffix = 'ccw'
+            else:
+                controller_suffix = 'pci'
+            bus_type = 'virtio-serial-%s' % controller_suffix
+
+        prefix = params.get('virtio_port_name_prefix')
+        name = prefix + serial_id if prefix else serial_id
+        params['chardev_name'] = name
+        chardev_id = 'chardev_%s' % serial_id
+        chardev_device = self.chardev_define_by_params(chardev_id, params, file_name)
+        serial_devices = self.serials_define_by_variables(serial_id,
+                                                          serial_type,
+                                                          chardev_id,
+                                                          bus_type=bus_type,
+                                                          serial_name=name,
+                                                          bus=params.get(
+                                                              "serial_bus"),
+                                                          nr=params.get(
+                                                              "serial_nr"),
+                                                          reg=params.get(
+                                                              "serial_reg"))
+
+        return [chardev_device] + serial_devices
+
+    def chardev_define_by_params(self, chardev_id, params, file_name=None):
+        """
+        Wrapper for creating -chardev device from params.
+        :param chardev_id: chardev id
+        :param params: chardev params
+        :param file_name: file name of chardev (optional)
+        :return: CharDevice object
+        """
+        backend = params.get('chardev_backend', 'unix_socket')
```

Comment:
No param `chardev_backend` now.
For serial, we have `serial_chardev_backend`,  if want to distinguish other device,  we may need make `backend` as a param of  `chardev_define_by_params` .